### PR TITLE
Training adjustments

### PR DESCRIPTION
**Pull request by LorenzLamm**
I tried to reproduce the performance of our best model, which is based on nnUNet. 
To this end, I added a lot of data augmentations, adjusted the model architecture, including deep supervision, and refined the training procedure (e.g. PolyLR scheduler, slight weight decay).
Now I'm happy with the performance. Please have a look and see if everything is fine :)

### CODE REVIEW

File: src/membrain_seg/dataloading/memseg_augmentation.py
Comment (LorenzLamm) [inline src/membrain_seg/dataloading/memseg_augmentation.py:1]: These augmentations are key to the network's performance in my opinion. 
I tried to reproduce nnUNet's data augmentations, which are mainly based on the batchgenerators package. I wanted to not rely on either nnunet or batchgenerators package, so I tried implementing all the augmentations with MONAI.
Many of the augmentations were already available, others I translated from batchgenerators to MONAI (see also dataloading.transforms.py) .

File: src/membrain_seg/dataloading/memseg_dataset.py
Comment (LorenzLamm) [inline src/membrain_seg/dataloading/memseg_dataset.py:1]: Just a classical Pytorch dataset with __getitem__ function that applies data augmentation transforms on each image separately. I also tried to do it batchwise in the collate_fn, but I think this doesn't really make sense, as it makes things more complicated and does not give a large time improvement.

File: src/membrain_seg/dataloading/memseg_pl_datamodule.py
Comment (LorenzLamm) [inline src/membrain_seg/dataloading/memseg_pl_datamodule.py:1]: Pytorch Lightning data module that initializes the datasets and data loaders.

File: src/membrain_seg/segment.py
Comment (LorenzLamm) [inline src/membrain_seg/segment.py:1]: That's the (very preliminary) prediction script. 
Most importantly, the parser takes as input the path to the tomogram to be segmented and the path to the model checkpoint that should be used.

Then the script performs the segmentation, also using 8-fold test time augmentation (using flipping along different axes). I experienced that this makes the segmentations more smooth and stable. I think it's worth doing it, even though it increases prediction speed by a lot.
(Was around 10-15 min / tomo in the end using an a100 GPU)

File: src/membrain_seg/train.py
Comment (LorenzLamm) [inline src/membrain_seg/train.py:1]: Training script: It initialized model checkpoint logging, as well as the loggers.

Note that this is still the WandB logger. For development, I like to keep this for now, but in the end, we should switch to Tensorboard logging (or even just CSV logging to avoid installing tensorflow?). But we'll just need to adjust one line for this.

Also some other stuff, like learning rate tracking is probably redundant here.

File: src/membrain_seg/training/optim_utils.py
Comment (LorenzLamm) [inline src/membrain_seg/training/optim_utils.py:1]: 1. Adjusted the DynUNet architecture to output feature maps from lower levels directly for deep supervision. Thus, we can compare the low-resolution outputs to downsampled versions of the GT map. (Instead of upsampling them first and then comparing to full-res GT). That's at least how nnunet does it -- not sure what's better here though.
2. Define Deep Supervision loss that computes masked DICE & Cross entropy loss for each downsampling level.
3. Define combination of Dice and CE loss that accepts ignore labels and does not evaluate the loss there.

File: src/membrain_seg/training/unet.py
Comment (LorenzLamm) [inline src/membrain_seg/training/unet.py:1]: Using the adjusted DynUNet (see optim_utils.py) with a lot of downsampling (5 time!) and same channel sizes as nnUNet. 
Logging masked DICE score + accuracy (latter only for sanity -- it's not very informative) and loss for both validation & training (epoch-wise).

File: src/membrain_seg/dataloading/data_utils.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    out_file = os.path.join(out_folder, os.path.basename(orig_data_path))
+    store_tomogram(out_file, predictions_np)
+    if store_probabilities:
+        predictions_np_thres = predictions.squeeze(0).squeeze(0).cpu().numpy() > 0.0
+        out_file_thres = os.path.join(
+            out_folder,
+            os.path.basename(orig_data_path)[:-4] + "_" + ckpt_token + "_segmented.mrc",
+        )
+        store_tomogram(out_file_thres, predictions_np_thres)
+    print("MemBrain has finished segmenting your tomogram.")
```
Comment (kevinyamauchi) [inline src/membrain_seg/dataloading/data_utils.py:45]: I think it would be nice to wrap print statements like this under some sort of `verbose` flag. For example, make `verbose` an input parameters and then

```python
if verbose:
    print....
```

File: src/membrain_seg/dataloading/memseg_augmentation.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        out_dir,
+        img,
+        label,
+        get_training_transforms(prob_to_one=True, return_as_list=True),
+    )
+    get_augmentation_timing(
+        imgs=[img, img],
+        labels=[label, label],
+        aug_sequence=get_training_transforms(prob_to_one=False, return_as_list=True),
+    )
```
Comment (kevinyamauchi) [inline src/membrain_seg/dataloading/memseg_augmentation.py:361]: Are uses expected to directly run this file or was this from testing? If it's just for testing, I think we should remove this. If not, I think we should replace the path variables to some sort of input parser as those paths won't work on the users system.

File: src/membrain_seg/dataloading/memseg_pl_datamodule.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+class MemBrainSegDataModule(pl.LightningDataModule):
+    """Pytorch Lightning datamodule for membrane segmentation."""
+
+    def __init__(self, data_dir, batch_size, num_workers, aug_prob_to_one=False):
+        super().__init__()
+        self.data_dir = data_dir
+        self.train_img_dir = os.path.join(self.data_dir, "imagesTr")
+        self.train_lab_dir = os.path.join(self.data_dir, "labelsTr")
+        self.val_img_dir = os.path.join(self.data_dir, "imagesVal")
+        self.val_lab_dir = os.path.join(self.data_dir, "labelsVal")
```
Comment (kevinyamauchi) [inline src/membrain_seg/dataloading/memseg_pl_datamodule.py:18]: If we are assuming a structure for the data directory, I think it would be nice if we added a note to the doctstring explaining the expected data directory structure.

File: src/membrain_seg/dataloading/transforms.py [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+        return data
+
+
+class SimulateLowResolutionTransform(Randomizable, Transform):
+    """
+    Batchgenerators' transform for simulating low resolution upscaling.
+
+    The image is downsampled using nearest neighbor interpolation using a
+    random scale factor. Afterwards, it is upsampled again with trilinear
+    interpolation, imitating the upscaling of low-resolution images.
```
Comment (kevinyamauchi) [inline src/membrain_seg/dataloading/transforms.py:161]: neat idea!

File: src/membrain_seg/dataloading/memseg_pl_datamodule.py
Comment (kevinyamauchi) [reply src/membrain_seg/dataloading/memseg_pl_datamodule.py:1]: good idea, thanks!

File: tests/membrain_seg/training/test_optim_utils.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import pytest
+
+
+@pytest.mark.filterwarnings("ignore::DeprecationWarning")
+def test_loss_fn_correctness():
+    from membrain_seg.training.optim_utils import test_loss_fn
+
+    test_loss_fn()
```
Comment (kevinyamauchi) [inline tests/membrain_seg/training/test_optim_utils.py:8]: Can you please move the test function here? While this works, I think it's better to separate the test code from the user-facing code. Also, it is helpful to test importing the functions here as they would be imported when the library is used.

**Review by kevinyamauchi — COMMENTED** [review]
Hey, @LorenzLamm ! This is great. Thanks for this. I have made some minor comments below. I think we merge soon so that we can keep moving.

Things I think we should change before merging:
- where possible, I think we should remove the `___main__` code from the various modules. I can see these are useful for testing, but I think we will want users to use our CLI (as we develop it) rather than calling the module files directly.
- Can we keep the `SemanticSegmentationUnet` in the `networks` module? Users/devs may want to import it for inference as well as training, so I think it would be more intuitive there. 

Additionally, I think there are a few things we can do in follow-up PRs
- convert print statements to logging that way users and developers can easily suppress them.
- add docstrings with parameters and return values explained

File: src/membrain_seg/training/optim_utils.py
Comment (kevinyamauchi) [reply src/membrain_seg/training/optim_utils.py:1]: Super nice!

File: src/membrain_seg/dataloading/memseg_augmentation.py (hunk 2 above)
Comment (LorenzLamm) [reply src/membrain_seg/dataloading/memseg_augmentation.py:361]: Ah yes. Probably makes sense to remove this.
I used this to see the effects of the data augmentations, but we should not require users to do it as well. So I'll just remove it.

File: src/membrain_seg/dataloading/memseg_pl_datamodule.py (hunk 3 above)
Comment (LorenzLamm) [reply src/membrain_seg/dataloading/memseg_pl_datamodule.py:18]: Good point. This is the data structure that I normally use for training, but maybe we should discuss at some point what data structure makes sense.
For now, I'll just add a docstring explaining the assumed data structure as it is right now.

File: tests/membrain_seg/training/test_optim_utils.py (hunk 5 above)
Comment (LorenzLamm) [reply tests/membrain_seg/training/test_optim_utils.py:8]: Yep, makes sense. Will do! :)

File: src/membrain_seg/dataloading/data_utils.py
Comment (LorenzLamm) [inline src/membrain_seg/dataloading/data_utils.py:1]: Corrected an error here: The script was saving score maps instead of segmentation masks if --store_probabilities flag was set to True.

File: src/membrain_seg/dataloading/memseg_augmentation.py
Comment (LorenzLamm) [inline src/membrain_seg/dataloading/memseg_augmentation.py:1]: I removed the testing functions for checking the effects of data augmentations

File: src/membrain_seg/dataloading/memseg_dataset.py
Comment (LorenzLamm) [inline src/membrain_seg/dataloading/memseg_dataset.py:1]: Removed the collate function (which was only used in a previous version with batch-wise augmentations).
Also removed the test functions for the dataset

File: src/membrain_seg/networks/unet.py
Comment (LorenzLamm) [inline src/membrain_seg/networks/unet.py:1]: Removed some redundant lines of code and renamed to /networks/unet.py again.

File: src/membrain_seg/training/optim_utils.py
Comment (LorenzLamm) [inline src/membrain_seg/training/optim_utils.py:1]: Moved the test functions to the tests folder instead of having them in the script itself.

**Comment by LorenzLamm** [comment]
> Hey, @LorenzLamm ! This is great. Thanks for this. I have made some minor comments below. I think we merge soon so that we can keep moving.
> 
> Things I think we should change before merging:
> 
> * where possible, I think we should remove the `___main__` code from the various modules. I can see these are useful for testing, but I think we will want users to use our CLI (as we develop it) rather than calling the module files directly.
> * Can we keep the `SemanticSegmentationUnet` in the `networks` module? Users/devs may want to import it for inference as well as training, so I think it would be more intuitive there.
> 
> Additionally, I think there are a few things we can do in follow-up PRs
> 
> * convert print statements to logging that way users and developers can easily suppress them.
> * add docstrings with parameters and return values explained

Thanks a lot for your feedback, Kevin! I tried to work in your suggested changes and removed most of the main() functions + changed the unet folder back to "networks". 
Also cleaned up some lines of code & removed some redundant functions.

As you suggest, probably it's best to replace the print statements and add more information to the docstring in later PRs. Need to invest some time for this I think :-D

**Review by kevinyamauchi — APPROVED** [review]
Looks good to me! Thank you for the quick turnaround, @LorenzLamm . I will merge this and make a couple of issues to track some of the things we are punting on.

Great work!
